File: sol/241/241.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        tokenized = []
        buff = []
        for c in expression:
            if c.isdigit():
                buff.append(c)
            else:
                tokenized.append(int(''.join(buff)))
                buff.clear()
                tokenized.append(c)
        tokenized.append(int(''.join(buff)))
        n = len(tokenized)
        memo = [[list() for __ in range(n)] for _ in range(n)]
        for i in range(0, n, 2):
            memo[i][i] = [tokenized[i]]
        for l in range(3, n + 1, 2):
            for i in range(0, n - l + 1, 2):
                j = i + l - 1
                for k in range(i + 1, j, 2):
                    for a in memo[i][k - 1]:
                        for b in memo[k + 1][j]:
                            if tokenized[k] == '+':
                                memo[i][j].append(a + b)
                            elif tokenized[k] == '-':
                                memo[i][j].append(a - b)
                            else:
                                memo[i][j].append(a * b)
        return memo[0][n - 1]

File: sol/241/241.cpp
struct Solution {
    vector<int> diffWaysToCompute(string &expression) {
        vector<char> tokenized;
        char buff = 0;
        for (char c : expression){
            if ('0' <= c && c <= '9'){
                buff *= 10;
                buff += c - '0';
            } else {
                tokenized.push_back(buff);
                buff = 0;
                tokenized.push_back(c);
            }
        }
        tokenized.push_back(buff);
        int n = tokenized.size();
        vector<vector<vector<int>>> memo(n, vector<vector<int>>(n));
        for (int i = 0; i < n; ++++i){
            memo[i][i].push_back(tokenized[i]);
        }
        for (int l = 3; l <= n; ++++l){
            for (int i = 0; i <= n - l; ++++i){
                int j = i + l - 1;
                for (int k = i + 1; k < j; ++++k){
                    for (int a : memo[i][k - 1]){
                        for (int b : memo[k + 1][j]){
                            switch (tokenized[k]){
                                case '+':
                                    memo[i][j].push_back(a + b);
                                    break;
                                case '-':
                                    memo[i][j].push_back(a - b);
                                    break;
                                case '*':
                                    memo[i][j].push_back(a * b);
                                    break;
                                default:
                                    break;
                            }
                        }
                    }
                }
            }
        }
        return memo[0][n - 1];
    }
};
